### PR TITLE
Add Alephium

### PR DESCRIPTION
**Pull request by h0ngcha0**

### CODE REVIEW

**Comment by nop33** [comment]
Hey @bumblefudge, a long time ago you [commented](https://github.com/alephium/alephium-frontend/issues/158#issuecomment-1845735619) that we could reach out to you once we are ready! This is a PR to integrate Alephium. Let us know if you have some time to check it out! Thanks a ton!

File: alephium/caip10.md
Comment (bumblefudge) [inline alephium/caip10.md:43]: This is interesting (and reminds me of a similar property of Polkadot addresses) but for the CASA readership you need to spell this out a little more: link to (and summarize as concisely as possible) _how_ you derive group membership from the address (and whether an address is specific to mainnet/devnet or not, and whether that too can be derived diagnostically from an address).

File: alephium/caip10.md
Comment (bumblefudge) [inline alephium/caip10.md:47]:
```suggestion
The native Alephium address encoding is a [Base58][]-encoded string of the type defined above.
[CAIP-10][] simply uses an entire native address as the `account_address` component.
```

Comment (bumblefudge) [reply alephium/caip10.md:47]: But this begs the question... what is the [chain_id segment](https://chainagnostic.org/CAIPs/caip-10#syntax) for each CAIP-10 example listed below in the test vectors?

You mention above that group membership can be derived from the address, but it was unclear how.  Do the address types correspond to group 0, 1, 2, and 3 respectively?

I was also unclear if each address "exists" only on a specific network (i.e., on mainnet or devnet), or if an address is equally valid on all 3, even if it only has history on 1 or 0 of them.

The CAIP-10 system is heirarchical and most CAIP-10 implementations have a usecase or two requiring them to know (if not a universal need to know) _which network to query for history_ of a given account/address, which is why there is a CAIP-2 embedded in each CAIP-10.
If this is not the case and addresses are universal, you might want to add a "-1" to the enum of networks you define in CAIP-2, like `offchain` or `universal`. This way, all CAIP-10s could be `alephium:universal/-1:{address}`.
Conversely, if there IS no universal/offchain address namespace in the app development model of alephium, do NOT introduce such a CAIP-2 wildcard but explain how a given address should be namespaced/mapped to CAIP-2s.

Fascinating stuff here, I love how different this is from other CAIP-2 and CAIP-10 profiles!

File: alephium/caip19.md
Comment (bumblefudge) [inline alephium/caip19.md:35]: Here too I'm confused about the missing `alephium:` prefix and CAIP-2 segment!

**Review by bumblefudge — CHANGES_REQUESTED** [review]
This is looking great so far, but the CAIP-10 and CAIP-19 profiles aren't quite conformant, and I am a little too pressed for time to do enough research to recommend more concretely what next steps would make for the best possible CAIP-2 segment to add to them!

File: alephium/caip10.md
Comment (h0ngcha0) [reply alephium/caip10.md]: Thanks for the great comment!

> You mention above that group membership can be derived from the address, but it was unclear how. Do the address types correspond to group 0, 1, 2, and 3 respectively?

I added a description in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/3afba3c2bfd832319e29de8af9950f59c81b16f0), basically the group information is encoded as part of the address and can be derived from it. 

Thanks for the context for `CAIP-2`. The address is indeed "universal" in the sense that it is valid for all networks, and the group information is implicitly embedded in the address. However dApps on Alephium can request an address from a certain group on a certain network. For example, a testnet dApp can request connection on chain id `alephium:testnet/1`, the wallet should connect the dApp with an address from group `1` on testnet. If a mainnet dApp request connection on chain `alephium:mainnet/-1`, then the wallet can connect the dApp with an address from any groups on mainnet.

So I am a bit unsure here as well, I specified chain id in`CAIP-2` as:
```
alephium + ":" + network_id + "/" + group
```
Because I thought dApp can use this format to request address from a certain group (or all groups) on a certain network, but since address is universal and group is already implicitly embedded in address, if feels like the address should be specified like `alephium:universal:{address}`. 

Really appreciate it if you could give us some advice here :pray:

File: alephium/caip19.md
Comment (h0ngcha0) [reply alephium/caip19.md:35]: Tokens are "universal" as well in the sense that they are not tied to a specified network. It should probably be something like `alephium:universal:{tokenId}`, but I am not sure about how to reconcile with the `CAIP-2` part in this case, similar to address.

**Comment by h0ngcha0** [comment]
> This is looking great so far, but the CAIP-10 and CAIP-19 profiles aren't quite conformant, and I am a little too pressed for time to do enough research to recommend more concretely what next steps would make for the best possible CAIP-2 segment to add to them!

Thanks a lot for your time! 

Do you think the following would work?

1. For chain id, we use `alephium:${network_id}/${group}`, because Alephium is a sharded blockchain that contains a number of smaller chains internally, so we want to be able to specify the specific "chain" inside of it through group. 

2. For address, we specify `alephium:${network_id}:${address}`. Because `address` already have the group information implicitly we don't need to specify it. 

What do you think? Any advice would be appreciated, thanks!

File: alephium/caip19.md
Comment (bumblefudge) [reply alephium/caip19.md:35]: yeah, that makes sense! if something is shard-specific or chain-specific it would go in the `chainId` reference (the second segment), and if you want to create a special case for `universal` addresses, you just add it to the caip2 profile "specification" section!

File: alephium/caip2.md
Comment (bumblefudge) [inline alephium/caip2.md:29]: I think what's missing here is an explanation of what a group _does_, and in what context "any group" would be meaningful.  for example: is there a 1:1 relationship between blockchains and groups? Does each group contain 4 blockchains, and each blockchain belong only to one group? 

It might also be worth mentioning _somewhere_, maybe in CAIP-19 or CAIP-10 or both, that smart-contracts are group-specific and that most wallets control an address in each group for this purpose, as [described here](https://docs.alephium.org/frequently-asked-questions/#why-can-the-dapp-only-connect-to-one-of-my-addresses) -- that's probably something crucial that multi-chain dapp or multichain-wallet developers will need a heads up about!

File: alephium/caip10.md
Comment (bumblefudge) [reply alephium/caip10.md]: Actually, I realized that `/` won't work in a CAIP-2 reference, the only diacritics allowed are `-` and `_`.  Since `-1` makes sense as the wildcard, that just leaves `_` as the discriminant.

Wouldn't it make more sense, given all this, to use this as the syntax?

`alephium + ":" + group { -1 | 0 | 1 | 2 | 3} + "_" + network_id `

**Comment by bumblefudge** [comment]
> 2. For address, we specify `alephium:${network_id}:${address}`. Because `address` already have the group information implicitly we don't need to specify it.

Actually, I would argue it's still worth encoding group in the CAIP-2 portion of a CAIP-10 because the group information requires quite a bit of sniffing, decoding and transformation to _extract_, whereas the counterparty handed a CAIP-10 may not have the compute to spare and would probably only do that much compute to _verify_ what they most often can just trust from the counterparty...

Lots of multichain stuff is kind of verbose or redundant in this way, to spare all parties in the system to keep all the tooling from every ecosystem on hand :D

File: alephium/caip19.md
Comment (h0ngcha0) [reply alephium/caip19.md]: Thanks! I added the special case for `universal` in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/d128daedf783e31eef823ee20259c85fd77977f4)

File: alephium/caip2.md
Comment (h0ngcha0) [reply alephium/caip2.md:42]: Thanks! Added some explanation according to your suggestion [here](https://github.com/ChainAgnostic/namespaces/pull/117/commits/d128daedf783e31eef823ee20259c85fd77977f4), do you think it is more clear this way?

File: alephium/caip10.md
Comment (h0ngcha0) [reply alephium/caip10.md]: Thanks! I have updated the chain id format in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/b16b4fe480accc5bde28f5134e70f9aabf0e9bb0)

**Comment by h0ngcha0** [comment]
> > 2. For address, we specify `alephium:${network_id}:${address}`. Because `address` already have the group information implicitly we don't need to specify it.
> 
> Actually, I would argue it's still worth encoding group in the CAIP-2 portion of a CAIP-10 because the group information requires quite a bit of sniffing, decoding and transformation to _extract_, whereas the counterparty handed a CAIP-10 may not have the compute to spare and would probably only do that much compute to _verify_ what they most often can just trust from the counterparty...
> 
> Lots of multichain stuff is kind of verbose or redundant in this way, to spare all parties in the system to keep all the tooling from every ecosystem on hand :D

Thanks, I have made the update in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/b16b4fe480accc5bde28f5134e70f9aabf0e9bb0)

**Review by bumblefudge — APPROVED** [review]
LGTM, waiting on 2nd editor approval or 2 independent approvals

**Comment by h0ngcha0** [comment]
> LGTM, waiting on 2nd editor approval or 2 independent approvals

Thanks a lot @bumblefudge for your time and suggestions!

Hey @obstropolos , would really appreciate it if you can take a look at this PR when you have time, thanks very much :pray:

File: alephium/README.md
Comment (obstropolos) [inline alephium/README.md:20]:
```suggestion
into groups, with separate chains responsible for processing
```

File: alephium/caip2.md
Comment (obstropolos) [inline alephium/caip2.md:24]:
```suggestion
from by addresses in group `G1` to addresses in group `G2` will be
```

**Review by obstropolos — CHANGES_REQUESTED** [review]
Small errata - will approve once corrections are made. 

Thanks for putting this together!

**Comment by h0ngcha0** [comment]
> Small errata - will approve once corrections are made.
> 
> Thanks for putting this together!

Both suggestions applied, thanks for your time!

**Review by obstropolos — APPROVED** [review]

**Pull request merged**
